### PR TITLE
Fix SIGSEGV when hook list rebuilt mid-iteration (issue #108)

### LINKED ISSUE #108: SIGSEGV in main_hook_function when plugin loads sub-modules during hook callback
## Summary

SIGSEGV in `main_hook_function` during `DispatchSpawn` when AMXX loads sub-modules at map start.

Reported by APG RoboCop: mm-p v1.21p105 + AMXX 1.10.5-APG on Debian 13 with ReHLDS (Counter-Strike).

## Root cause

AMXX loads metamod sub-modules from within its `pfnSpawn` hook (worldspawn). This calls `LOAD_PLUGIN` -> `plugin_addload` -> `load` -> `rebuild_hook_lists`. The rebuild does `realloc(hook_list_data, ...)` which can move the allocation. However, `main_hook_function` already holds a cached `plugs` pointer and `count` from before the plugin callback:

```c
list = Plugins->get_hook_list(api);
count = list->count;
plugs = list->plugs;
for(i=0; i < count; i++) {
    // plugs may now point to freed memory after realloc
}
```

After the plugin callback returns and the loop continues, `plugs` is a dangling pointer.

## Secondary issue

Even without the crash, a newly loaded plugin that registers a `pfnSpawn` hook will not be called for the current `DispatchSpawn` invocation, because the cached `count` was captured before the plugin was added.

## Backtrace

```
#0  main_hook_function() in metamod.so
#1  mm_DispatchSpawn(edict_s*) in metamod.so
#2  ED_LoadFromFile(char*) in engine_i486.so
#3  SV_LoadEntities() in engine_i486.so
#4  Host_Map() in engine_i486.so
...
#9  Host_InitializeGameDLL() in engine_i486.so
```

## Proposed fix

Replace dynamically allocated `hook_list_data` (realloc) with fixed-size arrays (`MPlugin *plugs[MAX_PLUGINS]` per list). Since MAX_PLUGINS is small, this is trivial memory. Then change `main_hook_function` to re-read `list->count` on each loop iteration instead of caching it locally. This solves both issues:

1. No crash: `plugs` pointer is stable (fixed array, never moves)
2. New plugins picked up: `rebuild_hook_lists` appends new plugins at end and increments count; the loop condition re-reads count each iteration and sees the new entry

The rebuild preserves ordering of existing plugins (iterates `plist` in index order), so no double-calls or skipped plugins.

### PR DESCRIPTION
**Pull request by jkivilin**
## Summary

- When a plugin callback triggers `rebuild_hook_lists()` (e.g. loading/unloading a plugin via AMXX), the cached `plugs` pointer and `count` in `main_hook_function` become stale because realloc moves the backing allocation, causing use-after-free crashes.
- Fix adds `hook_list_tables_updated` flag checked after each plugin call, with `MPluginList::find_plugin_after_rebuild()` using pointer comparison into `plist[]` to relocate the current plugin's index in the rebuilt list.
- Re-entrancy safe: uses `reentry_count` to track nesting depth. Only the outermost call clears the flag on entry/exit; inner calls capture it into a local `rebuild_happened` and propagate back on return, ensuring outer callers always detect rebuilds triggered during nested dispatch.
- Handles all edge cases: plugin added/removed/paused before/after current, self-removed, multiple simultaneous changes.

## Test plan

- [x] 26 unit tests for `find_plugin_after_rebuild` covering all combinations of load/unload/pause/unpause
- [x] 6 hook_list stability tests verifying stale pointer detection
- [x] 7 integration tests (`test_api_hook`) confirming no crash on real pre/post hook code paths
- [x] 5 re-entrancy tests verifying flag propagation across nested `main_hook_function` calls
- [x] Fork-based test confirming use-after-free crash without the fix (accepts SIGSEGV, SIGABRT, or ASan exit)
- [x] Full test suite passes: 169 mlist + 48 api_hook + all other test binaries
- [x] ASan/UBSan (`make sanitize`) passes clean
- [x] Tested on local rehlds server with amxmodx + jk_botti plugins

Closes #108

### CODE REVIEW

File: metamod/mlist.h [hunk 1]
```diff
@@ -73,6 +77,21 @@ class MPluginList : public class_metamod_new {
...
+			out_plugs = list->plugs;
+			out_count = list->count;
+			for(int j = 0; j < out_count; j++) {
+				if(out_plugs[j] == current_plugin)
+					return j;
+				if(out_plugs[j] > current_plugin)
+					return j - 1;
+			}
+			return out_count - 1;
+		}
```
[the remark this answers is not in the document]
Comment (jkivilin) [reply metamod/mlist.h:94]: If this happens, hook loop works the same way as it was before 'hook_table_lists' optimization. Not fixing as that would be API change.

File: metamod/api_hook.cpp [hunk 2]
```diff
@@ -134,6 +138,11 @@ static inline HOT_FUNC void DLLINTERNAL main_hook_function_void_t(unsigned int a
 		api_info.api_caller(pfn_routine, packed_args);
 		API_UNPAUSE_TSC_TRACKING();
 
+		if(unlikely(hook_list_tables_updated)) {
+			hook_list_tables_updated = mFALSE;
+			i = Plugins->find_plugin_after_rebuild(api, mFALSE, plug, count, plugs);
+		}
```
[the remark this answers is not in the document]
Comment (jkivilin) [reply metamod/api_hook.cpp]: Fixed. Added `reentry_count` to track nesting depth — only the outermost call clears the flag on entry/exit. Inner calls capture the flag into a local `rebuild_happened` and propagate it back on return. Tested with 5 new re-entrancy integration tests.

File: metamod/mlist.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (jkivilin) [reply metamod/mlist.h:94]: By design. `plist` is a fixed-size array member — pointer ordering is an inherent property that cannot change without redesigning the plugin storage. The 26 unit tests for this function serve as the invariant check.

File: metamod/mlist.cpp [hunk 3]
```diff
@@ -224,6 +226,8 @@ void DLLINTERNAL MPluginList::rebuild_hook_lists(void) {
 				hook_lists[api][1].plugs[idx[api][1]++] = &plist[i];
 		}
 	}
+
+	hook_list_tables_updated = mTRUE;
```
[the remark this answers is not in the document]
Comment (jkivilin) [reply metamod/mlist.cpp:230]: On realloc failure, `hook_list_data` is not freed — the old pointer remains valid. The hook lists are zeroed as a defensive measure (empty lists = no plugins called), but the in-flight loop's cached `plugs` pointer still points to valid allocated memory. Setting the flag here would cause the loop to call `find_plugin_after_rebuild` on zeroed lists, which would be worse. Current behavior is intentional.

File: metamod/tests/test_mlist.cpp [hunk 4]
```diff
@@ -4583,226 +4598,1764 @@ static int test_rebuild_endlist_bounds(void)
...
-	// trim_list
-	fail |= test_trim_list_empty();
-	fail |= test_trim_list_shrinks();
-	fail |= test_trim_list_shrinks_to_first();
+	int count;
+	MPlugin * const *plugs;
+	int new_i = list->find_plugin_after_rebuild(
+		e_api_dllapi, mFALSE, plugB, count, plugs);
+	ASSERT_INT(new_i, 1);
+	ASSERT_INT(count, 2);
```
[the remark this answers is not in the document]
Comment (jkivilin) [reply metamod/tests/test_mlist.cpp:4693]: These are out parameters unconditionally assigned by the callee before any read. The function has no early-return path. Not initializing them is fine and matches the calling pattern in api_hook.cpp.

File: metamod/tests/test_api_hook.cpp [hunk 5]
```diff
@@ -124,11 +144,57 @@ static int plugin2_pre_Spawn(edict_t *)
...
+
+	RETURN_META(MRES_IGNORED);
+}
+
 // ============================================================
 // Infrastructure: set up a MPluginList with mock plugins
 // ============================================================
 
-static MPluginList test_plugins("plugins.ini");
+MPluginList test_plugins("plugins.ini");
```
[the remark this answers is not in the document]
Comment (jkivilin) [reply metamod/tests/test_api_hook.cpp]: Fixed — restored `static`.

File: metamod/api_hook.cpp [hunk 6]
```diff
@@ -94,6 +95,7 @@ inline const api_info_t * DLLINTERNAL get_api_info(enum_api_t api, unsigned int
   #define MAYBE_META_DEBUG(do_debug, level, args) do { break; } while(0)
 #endif
 
+
```
[the remark this answers is not in the document]
Comment (jkivilin) [reply metamod/api_hook.cpp:99]: Removed in current revision.

File: metamod/tests/test_mlist.cpp [hunk 7]
```diff
@@ -19,13 +22,25 @@
...
+		return NULL;
+	if (ptr) {
+		size_t copy_size = size < last_realloc_old_size ? size : last_realloc_old_size;
+		memcpy(newptr, ptr, copy_size);
+		memset(ptr, 0, last_realloc_old_size);
+		free(ptr);
+	}
+	last_realloc_old_size = size;
+	return newptr;
 }
```
[the remark this answers is not in the document]
Comment (jkivilin) [reply metamod/tests/test_mlist.cpp:44]: Only `hook_list_data` is ever realloc'd in these test binaries. The single-global tracking is sufficient and verified by the full test suite passing under ASan.

File: metamod/tests/test_api_hook.cpp [hunk 8]
```diff
@@ -28,6 +31,24 @@
...
+		return NULL;
+	if (ptr) {
+		size_t copy_size = size < last_realloc_old_size ? size : last_realloc_old_size;
+		memcpy(newptr, ptr, copy_size);
+		memset(ptr, 0, last_realloc_old_size);
+		free(ptr);
+	}
+	last_realloc_old_size = size;
+	return newptr;
+}
```
[the remark this answers is not in the document]
Comment (jkivilin) [reply metamod/tests/test_api_hook.cpp:50]: Same as above — only one buffer is realloc'd per test binary. `malloc_usable_size` is not portable and returns the actual allocation size (often larger than requested), which would defeat the purpose of the test.

File: metamod/tests/test_api_hook.cpp [hunk 9]
```diff
@@ -124,11 +145,57 @@ static int plugin2_pre_Spawn(edict_t *)
...
+
+	RETURN_META(MRES_IGNORED);
+}
+
 // ============================================================
 // Infrastructure: set up a MPluginList with mock plugins
 // ============================================================
 
-static MPluginList test_plugins("plugins.ini");
+MPluginList test_plugins("plugins.ini");
```
[the remark this answers is not in the document]
Comment (jkivilin) [reply metamod/tests/test_api_hook.cpp]: Fixed — restored `static`.

File: metamod/mlist.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (jkivilin) [reply metamod/mlist.h:94]: Same as the earlier comment on this function — by design. `plist` is a fixed array, pointer ordering is structural.

File: metamod/api_hook.cpp [hunk 10]
```diff
@@ -113,18 +117,22 @@ static inline HOT_FUNC void DLLINTERNAL main_hook_function_void_t(unsigned int a
 
 	//Setup
 	status=MRES_UNSET;
+	if (likely(reentry_count++ == 0))
+		hook_list_tables_updated = mFALSE;
 
 	//Pre plugin functions
 	PublicMetaGlobals.prev_mres=MRES_UNSET;
+	rebuild_happened = hook_list_tables_updated;
```
[the remark this answers is not in the document]
Comment (jkivilin) [reply metamod/api_hook.cpp:125]: Acknowledged. The extra `find_plugin_after_rebuild` call on inner entry is harmless (finds current plugin at its correct index, no-op). Clearing the flag on inner entry would require save/restore logic that adds complexity for zero functional benefit — the current approach prioritizes correctness and simplicity over avoiding one redundant lookup in a rare re-entrancy+rebuild scenario.

File: metamod/api_hook.cpp [hunk 11]
```diff
@@ -94,6 +95,8 @@ inline const api_info_t * DLLINTERNAL get_api_info(enum_api_t api, unsigned int
   #define MAYBE_META_DEBUG(do_debug, level, args) do { break; } while(0)
 #endif
```
[the remark this answers is not in the document]
Comment (jkivilin) [reply metamod/api_hook.cpp:97]: HL engine is single-threaded by design (GoldSrc has no multi-threaded game DLL dispatch). This is a 20+ year old engine — not going to change.
